### DISCUSSION TITLE
Atomically read and update a counter

### DISCUSSION THREAD

**Discussion by emilianocalzada** [pocketbase/pocketbase#5484] [Q&A]
Hello, i'm using pocketbase as a backend for my application. Some months ago i asked how could i run database hooks in the background and the solution was to use go routines to make it a non-blocking function. It's working well

Inside my function im running this code:

```go
func extractDetailsForResult(app *pocketbase.PocketBase, record *models.Record) {
	// Do some stuff ...
	
	// Update the campaign object to show a progress bar (i have 2 fields total_results_count and processed_results_count)
	campaign := record.Get("campaign")
	campaignRecord, err := app.Dao().FindRecordById("campaigns", campaign.(string))
	if err != nil {
		fmt.Printf("Error finding campaign record: %v\n", err)
		return
	}
	processed_results_count := campaignRecord.Get("processed_results_count").(int)
	campaignRecord.Set("processed_results_count", processed_results_count+1)
	if err := app.Dao().SaveRecord(campaignRecord); err != nil {
		fmt.Printf("Error saving campaign record: %v\n", err)
		return
	}
	
app.OnModelAfterCreate("results").Add(func(e *core.ModelEvent) error {
	record := e.Model.(*models.Record)
	// Perform it inside a go routine for processing it in the background since it takes a bit too long for frontend to not update (better ux)
	go extractDetailsForResult(app, record)
	return nil
})
```

It works but i'm not sure if its safe. What will happen if two clients will increase this value at the same time? im running thousands go routines at the same time

Thank you, i really like the developing experience using pocketbase, so powerfull and flexible

**Comment by ganigeorgiev** [comment]
You can wrap the read and write in a [transaction](https://pocketbase.io/docs/go-records/#transaction):

```go
txErr := app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
    campaign := record.GetString("campaign")
    campaignRecord, err := txDao.FindRecordById("campaigns", campaign)
    if err != nil {
        app.Logger().Error("Error finding campaign record", "error", err, "campaign", campaign)
        return err
    }

    processed_results_count := campaignRecord.GetInt("processed_results_count")
    campaignRecord.Set("processed_results_count", processed_results_count+1)
    err = txDao.SaveRecord(campaignRecord)
    if err != nil {
        app.Logger().Error("Error saving campaign record", "error", err, "processed_results_count", processed_results_count+1)
        return err
    }

    return nil
})
```

_Side-note: this is not required if you have only one OnModelAfterCreate hook, but to avoid bugs in case in the future you decide to add another hook that changes `e.Model`, since you are running your code in a separate go routine it would be safer to pass a copy to your function with [`record.CleanCopy()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.20/models#Record.CleanCopy) (with the next release this is named just `Fresh()`). Additionally, I recommend using [`routine.FireAndForget(func() { ... })`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.20/tools/routine) since it auto recover in case it panics._
